### ISSUE TITLE
Enhancement: use Transifex for translations

### ISSUE THREAD

**Issue by Lars1963** [mrzapp/nextcloud-cookbook#46]
While the nextcloud-community is very active in translating, I suggest you use nextclouds transifex-account for the translation of the app. This would give your app a amazing boost. 🚀

See: https://help.nextcloud.com/t/translation-knowledge-valid-for-the-entire-nextcloud-project-wiki/51550

The guys from nextcloud will more as happy in helping you to setup your repo for transifex. 😁👍

**Comment by mrzapp** [comment]
@Lars as the time I get to spend on this app is very limited, I have actually been trying to contact the devs at NextCloud about moving this repo to their [organisation](https://github.com/nextcloud/), but I haven't gotten a response yet.

I bet the move to transifex would be a lot smoother if everybody didn't have to wait for me to have time to review pull requests. Do you have any idea how to move this along?

**Comment by christianlupus** [comment]
I just looked into this issue (closing some old ones). I think [this documentation](https://docs.nextcloud.com/server/13.0.0/developer_manual/app/l10n.html#setup-of-the-transifex-sync) is useful. If you want, I can create the `.tx/config` file but the adding of the "collaborator" bug needs to be done by @mrzapp.

I do not want to avoid this transition to transifex but the currently existing translations might or might not get converted smoothly. I am unsure what will happen, it might be necessary to contact the translation team directly.

**Comment by mrzapp** [comment]
@christianlupus not sure what you mean by "adding of the collaborator bug"?

**Comment by christianlupus** [comment]
@mrzapp The repository has a set of collaborators. These can only be changed by the repository owner (or administrator). For Transifex to work, you have to configure some settings (I will prepare a PR soon if you like), add a certain user to the list of collaborators. And finally registering the app with the main CI system.

There is a daily schedule that revisits all registered repositories and synchronizes the translations to and from transifex system. The changed i10n location files are then updated within the repository by the bot. Therefore the bot needs write access to the repository. I think [this file](https://github.com/nextcloud/docker-ci/blob/master/translations/handleAppsTranslations.sh) is run on the repository by the bot.

**Comment by christianlupus** [comment]
See here (from one of my repos):
![Screenshot_2020-03-16 christianlupus Test-PK-Pages](https://user-images.githubusercontent.com/8202922/76757533-07a9c080-6788-11ea-9282-74c72578b0c7.png)

**Comment by mrzapp** [comment]
@christianlupus OK, let's give it a spin! Please prepare the PR at your leisure :)

**Comment by christianlupus** [comment]
I just opened #160.

The branch to be synchronized with transifex has to be called `master`. (This is an unfortunate requirement by the CI system of nextcloud, see [here](https://github.com/nextcloud/docker-ci/blob/master/translations/handleAppTranslations.sh#L16-L27). The additional names `stable16` through `stable18` can be used at the moment.) This is merely an administrative issue for the cookbook app. I created a branch `master` for temporary usage that can be merged with `develop` if transifex is working.

Next steps required with admin rights:
1. Add protection rules to the important branches like `develop`, `master`, and `stable` such that no force-push is possible. Add more branches if you consider them critical.
1. Add the user @nextcloud-bot to the list of collaborators with write access to the repo.

Once this is set up, please give me a short hint. Then, I will enable the @nextcloud-bot to parse this repo as well. After this is done, we will see if further modifications are needed to conserve our current translations.

If everything is working all right, we can merge the branches and switch to `master` as development branch.

**Comment by mrzapp** [comment]
@christianlupus alright, I protected the branches and invited @nextcloud-bot now

**Comment by christianlupus** [comment]
I merged the PR to have a valid configuration.

**Comment by christianlupus** [comment]
I just opened [this PR](https://github.com/nextcloud/docker-ci/pull/209) to enable CI for the cookbook app. Lets see what happens after approval.

**Comment by skjnldsv** [comment]
Just merged, keep up the awesome work! :rocket: :hugs:

**Comment by skjnldsv** [comment]
> @christianlupus alright, I protected the branches and invited @nextcloud-bot now, awaiting its acceptance

cc @rullzer

**Comment by christianlupus** [comment]
@skjnldsv Thanks and sorry for the missing comma.

**Comment by christianlupus** [comment]
@mrzapp Was the invited bot accepted in the meantime? I registered on the transifex page but did not find any hints regarding the cookbook app. I wanted to check if the current translations are imported or not. I just want to ensure that it is not caused by missing rights.

**Comment by christianlupus** [comment]
Just an update on the progress:
The bot has parsed the repo and created a [subproject on transifex](https://www.transifex.com/nextcloud/nextcloud/cookbook/). At the moment, no translations were ported over to transifex. I opened an issue how to get the translations there.

**Comment by christianlupus** [comment]
OK, I joined the translation teams of all languages cookbook provides. I was able to manually import the `po` files. From the perspective of transifex I just added a bunch of translations.
To be honest, I do not know if it is necessary that the translations are reviewed or not. It might be a good idea, as it blocks the translations from been modified by anyone. But I do not yet know how this has to be done.

We will have to look at what happens as soon as the bot is working next time. I assume a new commit will be generated (assuming the contribution mail has been confirmed until then).

**Comment by rullzer** [comment]
So it seems l10n is ignored by your gitignore...

**Comment by christianlupus** [comment]
I am just preparing an update to `master`.

@rullzer I have one more question: Do we need the `pot` and `po` files to be tracked in the repo manually. As far as I understand things, `pot` are generated by the bot on the fly and the `po` are stored within transifex. Can you confirm this?

**Comment by rullzer** [comment]
> I am just preparing an update to `master`.
> 
> @rullzer I have one more question: Do we need the `pot` and `po` files to be tracked in the repo manually. As far as I understand things, `pot` are generated by the bot on the fly and the `po` are stored within transifex. Can you confirm this?

yeah those are extracted by the tool and pushed to transifex. No need to keep them

**Comment by christianlupus** [comment]
OK, let's see what is going to happen tonight. I removed all of the translation files except for some documentation stubs) and changed the `gitignore`.

**Comment by christianlupus** [comment]
@mrzapp Please keep in mind to update your deployment scripts.

**Comment by mrzapp** [comment]
@christianlupus so I'm still not 100% on what this all does. Should I remove the translation bits from the release script? I'm sorry this is such a dark area for me currently

**Comment by christianlupus** [comment]
@mrzapp I just checked what the state is. [This commit](https://github.com/mrzapp/nextcloud-cookbook/commit/ebca8d86185297ed72c398b55e6013ebf73a6a93) was generated on `master` branch. That is, Transifex integration is working in general. With the app.

How it works: Every night a bot on the Nextcloud CI implementation visits the cookbook app's repo. It scans the source code for localization strings. These are exported and pushed to Transifex. There any translator can work on the translations and put new translations.
Further any existing translations from Transifex are fetched by the bot. These are automatically converted into `*.json` and `*.js` in the `l10n` folder (see commit in `master`). Does this make sense for you now?

Now the next step would be to use `master` as the default development branch. This is the branch on which the translations are based. Any translation strings on this branch are pushed to Transifex so we should/must use it.
As soon as we switch to `master` branch, you must change the deploy scripts. Currently there is a call to `php /tmp/translationtool.phar convert-po-files` needed to convert the `po` files to the `js` and `json` files in `l10n`. This step is no longer needed. I changed the `draft-release.sh` script in the `master` branch already. But I am unsure, if this is the script you are using to deploy.

I close this issue now as I assume it to be done. If there are still unclear things, feel free to reopen. I'll try to help wherever possible.

**Comment by christianlupus** [comment]
PS: I just saw that there are yet some now languages from Transifex on our first day (I uploaded only the preexisting languages in the development branch). The rest was already translated on Transifex.

**Comment by skjnldsv** [comment]
Awesome! :)
Btw, is there any reason the default branch is develop and not master?
Which one provides the last releases?

If it is master I humbly suggest you set it back to develop to not confuse newcomers ;)

**Comment by skjnldsv** [comment]
Also, don't hesitate to do a shoutout to contributors doing an awesome job on translations! They deserve an incredible recognition and are unfortunately often forgotten!

**Comment by skjnldsv** [comment]
19 translated languages in 24h is an incredible performance :tada:

**Comment by christianlupus** [comment]
@skjnldsv This `master`/`development` issue was historically based. After the project was moved to Github (from Gitlab) there was a group of a few guys pushing together the app forward. This caused the need for a good git practice. Therefore we had the consent to have `develop` for the latest (but mostly stable) branch and `stable` for the branch on the app store. We might need to rename it simply `master` instead of `develop`.

The translation was really incredible. I added the previous 8 languages to Transifex but that more than doubled within 36 hours :tada:. I am not sure how to reach the translation contributors. I would like to thank them but this is now somehow separated from this repo....
